lib/models/users/matrix.js

```diff
@@ -75,13 +76,15 @@ MatrixUser.prototype.serialize = function() {
     this._data.localpart = this.localpart;
     return this._data;
 };
+
 /**
  * Make a userId conform to the matrix spec using QP escaping.
  * Grammar taken from: https://matrix.org/docs/spec/appendices.html#identifier-grammar
  */
 MatrixUser.prototype.escapeUserId = function() {
-    // Currently Matrix accepts / in the userId, although going forward it will be removed.
-    const badChars = new Set(this.localpart.replace(/([A-z0-9]|-|\.|=|_)+/g, ""));
+    // NOTE: Currently Matrix accepts / in the userId, although going forward it will be removed.
+    // NOTE: We also allow uppercase for the time being.
+    const badChars = new Set(this.localpart.replace(/([A-Z]|[a-z]|[0-9]|-|\.|=|_)+/g, ""));
```

Comment:
we should remove characters which aren't allowed, such as uppercase. I'm pretty sure Synapse has started failing to register uppercase usernames by now.